### PR TITLE
update porkbun api endpoint domain

### PR DESCRIPTION
**Pull request by cristian-aldea**
Addresses #751

### CODE REVIEW

**Review by alametti — APPROVED** [review]
Note per Porkbun's email:

**Deadline:** 2024-12-01 00:00:00 UTC

File: ddclient.in
Comment (dimo414) [inline ddclient.in:6731]: I suspect this link should not change; it works today but https://porkbun.com/api/json/v3/documentation is still what's linked from the Porkbun website footer so I'm guessing the documentation will stay on the main website when the API moves off it.

Comment (cristian-aldea) [reply ddclient.in:6731]: I think this documentation will also be moved, it's under the same `/api/json/v3/...` path that is used by the rest of the API.

On both https://porkbun.com/api/json/v3/documentation and https://api.porkbun.com/api/json/v3/documentation, you can see the following notice:

```
## API Hostname
api.porkbun.com

Please note that porkbun.com currently works and has historically been the correct hostname for our API. However, we will be migrating away from that hostname and the API will no function using it at some point in the future.
```

So https://porkbun.com/api/json/v3/documentation will no longer work at some point from my understanding

**Comment by cohenchris** [comment]
One quick question - how is the API endpoint handled by this plugin? Will the Porkbun endpoint need to be changed anywhere in the code itself? I could not find anywhere in this repository which explicitly hardcodes this endpoint, so I suspect we should be fine, but wanted to make sure before this PR is merged.

Thanks for the quick update on this issue!

**Comment by alametti** [comment]
> One quick question - how is the API endpoint handled by this plugin? Will the Porkbun endpoint need to be changed anywhere in the code itself? I could not find anywhere in this repository which explicitly hardcodes this endpoint, so I suspect we should be fine, but wanted to make sure before this PR is merged.

See the [file changed](https://github.com/ddclient/ddclient/pull/752/files) by this PR - lines 6826 and 6864 of `ddclient.in` currently hardcode the API endpoint as `url => "https://porkbun.com/api/json/v3/`.

**Comment by cohenchris** [comment]
> > One quick question - how is the API endpoint handled by this plugin? Will the Porkbun endpoint need to be changed anywhere in the code itself? I could not find anywhere in this repository which explicitly hardcodes this endpoint, so I suspect we should be fine, but wanted to make sure before this PR is merged.
> 
> See the [file changed](https://github.com/ddclient/ddclient/pull/752/files) by this PR - lines 6826 and 6864 of `ddclient.in` currently hardcode the API endpoint as `url => "https://porkbun.com/api/json/v3/`.

Thanks! I didn't scroll down 🙃

**Comment by TinfoilSubmarine** [comment]
Could we take the opportunity to make the endpoint configurable by the user like we have for other services? [I pushed a commit that makes this change](https://github.com/TinfoilSubmarine/ddclient/commit/e147878f12be3fd62221b7156f4e3da1ee0efe5c), you can just include it in this PR if you'd like.

**Comment by cristian-aldea** [comment]
Hello, any blockers for this CR? Just let me know what I need to do so a maintainer can merge it.

**Comment by MestreLion** [comment]
> If whoever needs to approve this CR to get it merged is okay with it, so am I. I'm not sure how useful it'll be unless porkbun decide to change api every so often which hopefully won't happen.

It is useful, as Porkbun also provides an alternate IPv4-only server at https://api-ipv4.porkbun.com . From their [API documentation](https://porkbun.com/api/json/v3/documentation):

> **IPv4 Only Hostname**
> Some folks prefer to force IPv4, especially when using the ping command to get back an IP address for dynamic DNS clients. The dedicated IPv4 hostname is **`api-ipv4.porkbun.com`**, use this instead of porkbun.com.

So a configurable hostname is very welcome in Porkbun protocol

**Comment by cristian-aldea** [comment]
> So a configurable hostname is very welcome in Porkbun protocol
Feel free to make a separate CR to include that change then. This CR is just meant to address #751.

It's been approved and not merged for 3 weeks now. @dimo414 @alametti what are the blockers to merging this?

**Comment by TinfoilSubmarine** [comment]
> > Could we take the opportunity to make the endpoint configurable by the user like we have for other services? [I pushed a commit that makes this change](https://github.com/TinfoilSubmarine/ddclient/commit/e147878f12be3fd62221b7156f4e3da1ee0efe5c), you can just include it in this PR if you'd like.
> 
> If whoever needs to approve this CR to get it merged is okay with it, so am I. I'm not sure how useful it'll be unless porkbun decide to change api every so often which hopefully won't happen.
> 
> EDIT: to include your commit, do I just need to cherry-pick it into my repo's main branch and push right?

Sorry, just saw your edit. Yep, you should be able to just cherry-pick e147878f12be3fd62221b7156f4e3da1ee0efe5c (once you've made your repo aware of my branch)

**Comment by cristian-aldea** [comment]
@TinfoilSubmarine ok done. Was able to add your fork as a remote and cherry-pick the commit.

@dimo414 @alametti let us know what blockers there are to merging this. It's been sitting for a while now.

**Comment by alametti** [comment]
> @dimo414 @alametti let us know what blockers there are to merging this. It's been sitting for a while now.

Nothing on my side. I'm not a maintainer - I just reviewed your changes and I agreed with them based on what Porkbun published.

**Comment by dimo414** [comment]
Not a maintainer either. I assume you're looking for a review from @rhansen

**Comment by cristian-aldea** [comment]
> Not a maintainer either. I assume you're looking for a review from @rhansen

Understood. Yeah I'd like a maintainer to review and merge. I wasn't sure who's who. @rhansen if you had time to review that would be nice.

**Comment by cohenchris** [comment]
Just as a note, I received a second reminder email from Porkbun to stop using this deprecated API. In that email, they've stated that the old API will be deprecated on 12/1

**Comment by Pyroseza** [comment]
here's the 2nd email from Porkbun team:

> Hello there,
> 
> The reason you are receiving this email is because you have an API key associated with your account and are sending commands to a deprecated API hostname. We sent a previous warning on 2024-10-11. In order to ensure that any apps or tools you may have that utilize our API continue to work, we wanted to let you know about some upcoming critical updates.
> 
> We know these kinds of changes are annoying but the time has come to separate the processing of incoming API commands from our website. In previous documentation the hostname [porkbun.com](http://porkbun.com/) was used for the API but we have recently updated it to [api.porkbun.com](http://api.porkbun.com/) and in the near future will start enforcing its use. Please update the hostname for our API to [api.porkbun.com](http://api.porkbun.com/) as soon as possible.
> 
> CRITICAL UPDATE DETAILS
> 
> Type: API Hostname Change
> 
> Old Value: [porkbun.com](http://porkbun.com/)
> 
> New Value: [api.porkbun.com](http://api.porkbun.com/)
> 
> Deadline: 2024-12-01 00:00:00 UTC
> 
> Please note that after the deadline, API calls made to the old hostname will no longer be allowed. If you have any questions or concerns please contact support.
> 
> 
> Thank you,
> The Porkbun Team

**Comment by cristian-aldea** [comment]
I see that last PR was merged by @rhansen. @rhansen do you have time to look at this? porkbun provider will break in a month otherwise

File: ddclient.in
Comment (andyporkbun) [reply ddclient.in:6732]: Just double-checked with our dev and can confirm the docs will remain hosted at Porkbun.com, so https://porkbun.com/api/json/v3/documentation will be the correct link for the foreseeable future.

Comment (cristian-aldea) [reply ddclient.in:6732]: They are also hosted on api.porkbun.com. Both links would be correct then. Might as well use the new one imo.

Comment (cristian-aldea) [reply ddclient.in:6732]: Just checked the new link and the css isn't working. Maybe they aren't supporting it yet. Either way doesn't make a different to me. I'll change it back to https://porkbun.com/api/json/v3/documentation

Comment (cristian-aldea) [reply ddclient.in]: Also, just going to mention real quick, I just saw #757 now and I didn't know you worked at porkbun. thanks for chiming in. I thought the api doc would move, but that's not the case.

**Comment by AhnHEL** [comment]
Is the Maintainer still actively involved?  Are we back to ddclient being unmaintained again?

**Comment by kellerassel007** [comment]
Please merge this PR, we need this API update until deadline!

**Comment by AD-Wright** [comment]
Please remain calm, we still have two weeks before the deadline!  Sure, it would be great to have it sooner for patching and testing purposes, however the maintainer may only be able to check-in once a week or so.

**Comment by rageworx** [comment]
So I ... been wait this merge to be granted soon, but maintainer still need time for something verified or need to be checked for Porkbun API changes ?
Hope this update be merged to main ASAP until deadline so far ...

**Comment by TinfoilSubmarine** [comment]
For those getting anxious about this, note that the "build" process for this is pretty easy since it's just a Perl script. You'll need to download the archive from this branch (https://github.com/cristian-aldea/ddclient/archive/b6be379c9322098e446bea21f610054b9473ba70.zip), and then you can [follow the "Manual Installation" instructions in the README](https://github.com/ddclient/ddclient?tab=readme-ov-file#manual-installation) (remember to run `./autogen`!)

**Comment by ajhall** [comment]
There are also some other DDNS updaters that have already updated their Porkbun support:
- https://github.com/qdm12/ddns-updater
- https://github.com/mietzen/porkbun-ddns

**Comment by alametti** [comment]
> For those getting anxious about this, note that the "build" process for this is pretty easy since it's just a Perl script. You'll need to download the archive from this branch (https://github.com/cristian-aldea/ddclient/archive/b6be379c9322098e446bea21f610054b9473ba70.zip), and then you can [follow the "Manual Installation" instructions in the README](https://github.com/ddclient/ddclient?tab=readme-ov-file#manual-installation) (remember to run `./autogen`!)

And make sure to back up your current `/etc/ddclient/ddclient.conf` (or whatever your configuration file is) to somewhere outside that file path (if using defaults) - otherwise, the manual installation may overwrite it.

I just built it following the Manual installation instructions after cloning [the fork for this PR](https://github.com/cristian-aldea/ddclient/tree/master) and everything seems to be working.

**Comment by AD-Wright** [comment]
Once ddclient is updated, what would the procedure be to switch back?  Backup the configuration files and reinstall?

**Comment by AD-Wright** [comment]
> Note per Porkbun's email:
> 
> **Deadline:** 2024-12-01 00:00:00 UTC

Deadline has now passed.  If you are coming here because you use ddclient, and it is now broken, see the comments above from ajhall and alametti.  Compiling a branch with this fix and installing it fairly straightforward (although this will then leave you on a dead branch that won't be getting further updates).

**Comment by luisnabais** [comment]
This is the easiest of fixes and yet it's months since the warning. I don't believe we can trust the development of this tool. I think I'll change to a different tool for this job.

**Comment by cristian-aldea** [comment]
> * https://github.com/qdm12/ddns-updater

Yeah I've switched to https://github.com/qdm12/ddns-updater and it's been working well for me. I recommend it to anyone reading this thread and waiting on the fix for ddclient.

I hope we hear from the maintainer soon. But I know maintaining open source software is hard so they might be taking a break or something.

The one time I decided to contribute to OS and it gets stuck waiting on approvals 😭

**Comment by rhansen** [comment]
Apologies for taking so long.  I was hoping some of the other maintainers would step in while I was away, looks like that didn't happen.  :(

**Pull request merged**
